contracts/src/ocr2/aggregator_proxy.cairo

```diff
@@ -0,0 +1,286 @@
+
+    use starknet::ContractAddress;
+    use starknet::ContractAddressIntoFelt252;
+    use starknet::ContractAddressZeroable;
+    use starknet::Felt252TryIntoContractAddress;
+    use integer::Felt252TryIntoU128;
+    use starknet::StorageAccess;
+    use starknet::StorageBaseAddress;
+    use starknet::SyscallResult;
+    use integer::U128IntoFelt252;
+    use integer::U128sFromFelt252Result;
+    use starknet::storage_read_syscall;
+    use starknet::storage_write_syscall;
+    use starknet::storage_address_from_base_and_offset;
+    use integer::u128s_from_felt252;
+    use option::OptionTrait;
+    use traits::Into;
+    use traits::TryInto;
+    use zeroable::Zeroable;
+
+    use chainlink::ocr2::aggregator::IAggregator;
+    use chainlink::ocr2::aggregator::Round;
+    use chainlink::libraries::ownable::Ownable;
+    use chainlink::libraries::simple_read_access_controller::SimpleReadAccessController;
+
+    const SHIFT: felt252 = 0x100000000000000000000000000000000_felt252;
+    const MAX_ID: felt252 = 0xffffffffffffffffffffffffffffffff_felt252;
+
+    #[derive(Copy, Drop, Serde)]
+    struct Phase {
+        id: u128,
+        aggregator: ContractAddress
+    }
+
+    impl PhaseStorageAccess of StorageAccess::<Phase> {
+        fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult::<Phase> {
+            Result::Ok(
+                Phase {
+                    id: storage_read_syscall(
+                        address_domain, storage_address_from_base_and_offset(base, 0_u8)
+                    )?.try_into().unwrap(),
+                    aggregator: storage_read_syscall(
+                        address_domain, storage_address_from_base_and_offset(base, 1_u8)
+                    )?.try_into().unwrap(),
+                }
+            )
+        }
+
+        fn write(address_domain: u32, base: StorageBaseAddress, value: Phase) -> SyscallResult::<()> {
+            storage_write_syscall(
+                address_domain, storage_address_from_base_and_offset(base, 0_u8), value.id.into()
+            )?;
+            storage_write_syscall(
+                address_domain, storage_address_from_base_and_offset(base, 1_u8), value.aggregator.into()
+            )
+        }
+    }
+
+    struct Storage {
+        _current_phase: Phase,
+        _proposed_aggregator: ContractAddress,
+        _phases: LegacyMap<u128, ContractAddress>
+    }
+
+    #[event]
+    fn AggregatorProposed(current: ContractAddress, proposed: ContractAddress) {}
+
+    #[event]
+    fn AggregatorConfirmed(previous: ContractAddress, latest: ContractAddress) {}
+
+    impl AggregatorProxy of IAggregatorProxy {
+        fn latest_round_data() -> Round {
+            let phase = _current_phase::read();
+            let aggregator = IAggregatorDispatcher { contract_address: phase.aggregator };
+            let round = aggregator.latest_round_data();
+            Round {
+                round_id: (phase.id.into() * SHIFT) + round.round_id,
```

Comment:
so higher 128 bits are for phase id and lower 124 bits for round id?

Reply:
The other way around actually, upper 124 are phase_id (since we will never need that many) and round_id is a full u128